fe1-web/network/NetworkConnection.ts

```diff
@@ -0,0 +1,155 @@
+
+  constructor(address: string) {
+    const ws = new W3CWebSocket(address);
+
+    ws.onopen = () => { console.info(`Initiating web socket : ${address}`); };
+    ws.onmessage = (message: any) => {
+      console.log(`Received a new message from '${address}' : `, message.data);
+
+      let response: JsonRpcResponse;
+      try {
+        response = JsonRpcResponse.fromJson(message.data);
+      } catch (e) {
+        throw new ProtocolError('Answer is not valid Json');
+      }
+
+      if (response.id !== UNDEFINED_ID) {
+        const pendingResponse = this.payloadPending.get(response.id);
+        if (pendingResponse === undefined) {
+          throw new NetworkError(`Received a response which id = ${response.id} does not match any pending requests`);
+        }
+
+        this.payloadPending.delete(response.id);
+
+        pendingResponse.resolvePromise(response);
+        this.onMessageHandler(response);
+      }
+    };
+    ws.onclose = () => { console.info(`Closed websocket connection : ${address}`); };
+    ws.onerror = (event: any) => { console.error(`WebSocket error observed on '${address}' : `, event); };
+
+    this.address = address;
+    this.ws = ws;
+  }
+
+  public setResponseHandler(handler: ResponseHandler) {
+    this.onMessageHandler = handler;
+  }
+
+  public sendPayload(payload: JsonRpcRequest): Promise<JsonRpcResponse> {
+    // Check that the websocket connection is ready
+    if (!this.ws.readyState) {
+      return new Promise(() => {
+        this.waitWebsocketReady().then(() => this.sendPayload(payload));
+      });
+    }
+
+    // websocket ready to be used, message can be sent
+    let query: JsonRpcRequest = payload;
+
+    // Note: this only works because react/Js is single threaded
+    query = new JsonRpcRequest({
+      ...payload,
+      id: this.payloadSentCount,
+    });
+    this.payloadSentCount += 1;
+
+    const promise = new Promise<JsonRpcResponse>((resolve, reject) => {
+      const timeoutId = setTimeout(
+        () => reject(new NetworkError(
+          `Maximum waiting time of ${WEBSOCKET_MESSAGE_TIMEOUT_MS} ms reached, dropping query`,
+        )), // FIXME if reject, delete from queue
+        WEBSOCKET_MESSAGE_TIMEOUT_MS,
+      );
+
+      const pendingResponse: PendingResponse = {
+        promise: promise,
+        resolvePromise: resolve,
+        rejectPromise: reject,
+        timeoutId: timeoutId,
+      };
+
+      this.payloadPending.set(
+        query.id as number, // as this point, there will be a number as id
+        pendingResponse,
+      );
+
+      console.log(`Sending this message on '${this.address}' : `, query);
+      this.ws.send(JSON.stringify(query));
+    });
+
+    return promise;
+  }
+
+  public disconnect(): void {
+    this.ws.close();
+  }
+
+  private waitWebsocketReady(): Promise<void> {
+    if (this.ws.readyState) {
+      return new Promise((resolve) => { resolve(); });
```

Comment:
This time I think you can really do this :) 
```suggestion
      return Promise.resolve();
```